nixos/modules/services/continuous-integration/buildbot-master/default.nix

```diff
@@ -0,0 +1,127 @@
+      };
+
+      extraGroups = mkOption {
+        type = types.listOf types.str;
+        default = [ "nixbld" ];
+        description = "List of extra groups that the buildbot user should be a part of.";
+      };
+
+      home = mkOption {
+        default = "/home/buildbot";
+        type = types.path;
+        description = "Buildbot home directory.";
+      };
+
+      buildbotDir = mkOption {
+        default = "${cfg.home}/master";
+        type = types.str;
+        description = "Specifies the Buildbot directory.";
+      };
+
+      listenAddress = mkOption {
+        default = "0.0.0.0";
+        type = types.str;
+        description = "Specifies the bind address on which the buildbot HTTP interface listens.";
+      };
+
+      port = mkOption {
+        default = 8010;
+        type = types.int;
+        description = "Specifies port number on which the buildbot HTTP interface listens.";
+      };
+
+      packages = mkOption {
+        default = [ pkgs.stdenv pkgs.buildbot-ui ]; # config.programs.ssh.package pkgs.nix ];
+        type = types.listOf types.package;
+        description = "Packages to add to PATH for the buildbot process.";
+      };
+
+      environment = mkOption {
+        default = {};
+        type = with types; attrsOf str;
+        description = "Additional environment variables to be passed.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.extraGroups = optional (cfg.group == "buildbot") {
+      name = "buildbot";
+    };
+
+    users.extraUsers = optional (cfg.user == "buildbot") {
+      name = "buildbot";
+      description = "buildbot user";
+      isNormalUser = true;
+      createHome = true;
+      home = cfg.home;
+      group = cfg.group;
+      extraGroups = cfg.extraGroups;
+      useDefaultShell = true;
+    };
+
+    systemd.services.buildbot-master = {
+      description = "Buildbot Continuous Integration Server";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment =
+        let
+          selectedSessionVars =
+            lib.filterAttrs (n: v: builtins.elem n [ "NIX_PATH" ])
+              config.environment.sessionVariables;
+        in
+          selectedSessionVars //
+            { BUILDBOT_HOME = cfg.home;
+              BUILDBOT_DIR = cfg.buildbotDir;
+              NIX_REMOTE = "daemon";
+            } //
+            cfg.environment;
+
+      serviceConfig = {
+        Type = "forking";
+        User = cfg.user;
+        Group = cfg.group;
+        WorkingDirectory = cfg.home;
+      };
+
+      path = cfg.packages;
+      preStart = ''
+        ${pkgs.buildbot-ui}/bin/buildbot create-master ${cfg.buildbotDir}
+        sed -i '/www/,+1 d' ${cfg.buildbotDir}/master.cfg.sample
+        mv -v ${cfg.buildbotDir}/master.cfg.sample ${cfg.buildbotDir}/master.cfg
+        echo "c['www'] = dict(port=8010)" >> ${cfg.buildbotDir}/master.cfg
+        cat ${cfg.buildbotDir}/master.cfg
```

Comment:
It seems like it would be nice for master.cfg to exist in the store, to avoid the config from changing without Nix being aware of it (since it'll just overwrite it anyway).

Maybe use `-c` to `create-master` to point it at a master.cfg in the store?